Tests for ShuffleDataset

feihugis opened the pull request:
This PR adds tests for `ShuffleDataset`:

- https://github.com/tensorflow/tensorflow/commit/481d6d75f6c792476b902d67b029a644eb7c94bf: add `resource_mgr` to `op_kernel_ctx` and `iterator_ctx`; `resource_mgr` is needed for `ShuffleDataset` to generate the random seed.

- https://github.com/tensorflow/tensorflow/commit/54d862dbbf1b0469b5df1a8246ef743c29a6f66b: tests for `ShuffleDataset`.

cc: @jsimsa

feihugis committed:
Add resource_mgr to the op_kernel_ctx and iterator_ctx in DatasetOpsTestBase

feihugis committed:
Tests for ShuffleDatasetOp

jsimsa commented on tensorflow/core/kernels/data/shuffle_dataset_op_test.cc:
```diff
@@ -0,0 +1,665 @@
+  return tensors;
+}
+
+// Test case 1: normal case with reshuffle_each_iteration = false
```
Do you have any test that checks that reshuffle each iteration behaves as expected?

feihugis replied:
Will add the tests to check the reshuffle behaviors.

feihugis replied:
fixed

jsimsa commented on tensorflow/core/kernels/data/shuffle_dataset_op_test.cc:
```diff
@@ -0,0 +1,665 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Do you plan to add tests for the ShuffleAndRepeat op in a separate PR?

feihugis replied:
Yes, I was planning to add another file `shuffle_and_repeat_dataset_op_test.cc` to test for the ShuffleAndRepeat op in a separate PR. But it looks like that it will share a lot of code with this PR, it may be better to add the tests for `ShuffleAndRepeat` in the same file. Which way do you prefer to?

jsimsa replied:
SGTM

feihugis replied:
fixed via https://github.com/tensorflow/tensorflow/pull/27969/commits/bccf25812b0dbe50aaa06d06b5c8c1c22dd9c371

jsimsa commented on tensorflow/core/kernels/data/shuffle_dataset_op_test.cc:
```diff
@@ -0,0 +1,665 @@
+  // order, so we need to consider the element sequence when evaluating the
+  // result for this case.
+  bool expect_items_equal = test_case.buffer_size.flat<int64>()(0) > 1;
+  TF_EXPECT_OK(ExpectEqual(out_tensors, test_case.expected_outputs,
```
The shuffle order should be fixed given the seeds so we should be able to check for equality. 

On a related note, the `ExpectEqual` method is misleading. If it is actually not testing for equality (`expect_items_equal == false`), then it is not clear what it is doing? My suggestion would be to get rid of the boolean argument and make sure that all vector comparisons test for equality.

feihugis replied:
Yes, we can check the equality with considering the element order.

Do you mean to change the above code to be as below
```
TF_EXPECT_OK(ExpectEqual(out_tensors, test_case.expected_outputs,
                           /*expect_items_equal*/ false));
```

Or do you mean to change the API of the utility function [`ExpectEqual`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/dataset_test_base.h#L54-L59) to remove the `expect_items_equal` argument?

The utility function [`ExpectEqual`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/dataset_test_base.h#L54-L59) wants to work like the python utility function [`assertDatasetProduces`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/test_base.py#L138). When `expect_items_equal  == false`, it will return true if the two vectors have the same elements and same order; When `expect_items_equal == true`, it will return true if the two vectors have same elements (no need to have same order). The argument name `expect_items_equal` is confusing. How about changing it to be `regardless_of_order`?

jsimsa replied:
Rename the argument to `compare_order`.

feihugis replied:
fixed via https://github.com/tensorflow/tensorflow/pull/27969/commits/6238de4711c8a0e5d25394374a1a32e5f5a8755d

jsimsa commented on tensorflow/core/kernels/data/shuffle_dataset_op_test.cc:
```diff
@@ -0,0 +1,665 @@
+  // order, so we need to consider the element sequence when evaluating the
+  // result for this case.
+  bool expect_items_equal = test_case.buffer_size.flat<int64>()(0) > 1;
+  TF_EXPECT_OK(ExpectEqual(out_tensors, test_case.expected_outputs,
```
Same comment, we should be testing for equality for all test cases.

feihugis committed:
Refactor ExpectEqual() function
1. Rename the argument name from `expect_items_equal` to `compare_order`
2. Enable the function to return tensorflow::errors::Internal if the two input tensor (vectors) are not equal.

feihugis committed:
Improve the Cardinality function and validate the input count

feihugis committed:
Enhance ShuffleDataset tests and add tests for ShuffleAndRepeatDataset

feihugis commented:
@jsimsa Thanks for your review! The comments are addressed by the following commits:

- https://github.com/tensorflow/tensorflow/pull/27969/commits/6238de4711c8a0e5d25394374a1a32e5f5a8755d: rename `expect_items_equal` to `compare_order`; also, enable `ExpectEqual()` return the error status if the input tensors are not equal.

- https://github.com/tensorflow/tensorflow/pull/27969/commits/8ca3f0c2088e8c30d9c3236c98f053aa2e45c0d0: Improve the cardinality function and check the input count

- https://github.com/tensorflow/tensorflow/pull/27969/commits/bccf25812b0dbe50aaa06d06b5c8c1c22dd9c371: add the reshuffle tests and tests for `ShuffleAndRepeatDataset`

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -52,11 +52,11 @@ class DatasetOpsTestBase : public ::testing::Test {
 
   // The method validates whether the two tensor vectors have the same tensors.
-  // If `expect_items_equal` is true, the method will only evaluate the two
+  // If `compare_order` is false, the method will only evaluate the two
```
s/evaluate the/evaluate whether the/

feihugis replied:
It is fixed via https://github.com/tensorflow/tensorflow/pull/27969/commits/9ea146a83b5ea9d2d1dceee54ff2994e36b1335b

jsimsa reviewed (dismissed):
Thanks @feihugis .

feihugis committed:
Revise the docstring for ExpectEqual()

feihugis commented:
@jsimsa 8 of the tests fail, but could not access the logs. I rebase my code to be the latest and run the tests for tf.data on my laptop using `bazel test //tensorflow/core/kernels/data:*`. They can run successfully.

I just added another commit (https://github.com/tensorflow/tensorflow/pull/27969/commits/9ea146a83b5ea9d2d1dceee54ff2994e36b1335b) to revise the docstring. Could you please reapprove it to trigger the tests?

feihugis commented:
@jsimsa Just a kindly reminder that this commit (https://github.com/tensorflow/tensorflow/commit/9ea146a83b5ea9d2d1dceee54ff2994e36b1335b) is submitted to address [your comment](https://github.com/tensorflow/tensorflow/pull/27969#discussion_r277811524). Could you help review the change when you get a chance?

feihugis commented:
`Ubuntu Sanity` test failed, but the log is not accessible. All the changed files have been checked by `clang-format`/`buildifier` command. Did not see any issue at my side.
